### ISSUE TITLE
Subtype and callargs validation for `functools.singledispatch`

### ISSUE THREAD

**Issue by iyume** [microsoft/pyright#4472]
**Is your feature request related to a problem? Please describe.**

More support for singledispatch:

1. first parameter subtype check
2. call args check

**Describe the solution you'd like**
```python
from typing import Union

from functools import singledispatch


@singledispatch
def foo(tp: Union[int, str], verbose: bool) -> int:
    ...


@foo.register
def foo_int(tp: int) -> int:  # no parameter check for verbose
    ...


@foo.register
def foo_error(tp: complex) -> int:
    # mypy: Dispatch type "complex" must be subtype of fallback function first argument "Union[int, str]"
    ...


foo(1)  # mypy: Missing positional argument "verbose" in call to "foo"

```

**Comment by erictraut** [comment]
The typeshed `functools.pyi` stubs are incomplete for `singledispatch` and `singledispatchmethod`. They should make use of `ParamSpec`. They predate `ParamSpec`, and it appears that no one has updated them accordingly. Now that mypy has support for `ParamSpec`, typeshed stubs are starting to make use of them. Please file a feature request and/or PR with the typeshed project.

I'm not an expert on `singledispatch` or `singledispatchmethod`, but here's a first attempt at how one might define them using a `ParamSpec`.

```python
class _SingleDispatchCallable(Generic[_S, _P, _T]):
    registry: types.MappingProxyType[Any, Callable[_P, _T]]
    def dispatch(self, cls: Any) -> Callable[Concatenate[Any, _P], _T]: ...
    # @fun.register(complex)
    # def _(arg, verbose=False): ...
    @overload
    def register(
        self, cls: type[Any], func: None = ...
    ) -> Callable[
        [Callable[Concatenate[Any, _P], _T]], Callable[Concatenate[Any, _P], _T]
    ]: ...
    # @fun.register
    # def _(arg: int, verbose=False):
    @overload
    def register(
        self, cls: Callable[Concatenate[Any, _P], _T], func: None = ...
    ) -> Callable[Concatenate[Any, _P], _T]: ...
    # fun.register(int, lambda x: x)
    @overload
    def register(
        self, cls: type[Any], func: Callable[Concatenate[Any, _P], _T]
    ) -> Callable[Concatenate[Any, _P], _T]: ...
    def _clear_cache(self) -> None: ...
    def __call__(__self, __val: _S, *args: _P.args, **kwargs: _P.kwargs) -> _T: ...

def singledispatch(
    func: Callable[Concatenate[_S, _P], _T]
) -> _SingleDispatchCallable[_S, _P, _T]: ...

if sys.version_info >= (3, 8):
    class singledispatchmethod(Generic[_S, _P, _T]):
        dispatcher: _SingleDispatchCallable[_S, _P, _T]
        func: Callable[Concatenate[Any, _S, _P], _T]
        def __init__(self, func: Callable[Concatenate[Any, _S, _P], _T]) -> None: ...
        @property
        def __isabstractmethod__(self) -> bool: ...
        @overload
        def register(
            self, cls: type[Any], method: None = ...
        ) -> Callable[
            [Callable[Concatenate[Any, Any, _P], _T]],
            Callable[Concatenate[Any, Any, _P], _T],
        ]: ...
        @overload
        def register(
            self, cls: Callable[Concatenate[Any, Any, _P], _T], method: None = ...
        ) -> Callable[Concatenate[Any, Any, _P], _T]: ...
        @overload
        def register(
            self, cls: type[Any], method: Callable[Concatenate[Any, Any, _P], _T]
        ) -> Callable[Concatenate[Any, Any, _P], _T]: ...
        def __get__(
            self, obj: _S, cls: type[_S] | None = ...
        ) -> Callable[Concatenate[Any, _S, _P], _T]: ...
```

With these changes, pyright properly detects missing parameters for registered calls and missing arguments for an invocation of the single dispatch function.

These changes don't allow pyright to detect the case where first parameter of a register call is not a subtype of the fallback function. That would require special casing in the type checker. I generally don't like to add special-case logic like that, but we have made some exceptions in the past if there is sufficient demand from pyright users.

I recommend starting with the typeshed changes. That will get you most of what you're looking for, and these changes will benefit other type checkers as well.

**Comment by iyume** [comment]
Thanks for workaround. It looks correct and good. I'll ask the people of typeshed.
